shared/api/chat_template.cc

```diff
@@ -304,6 +304,59 @@ static json NormalizeTools(const char* tools_str) {
   return normalized;
 }
 
+/*
+ * This function normalizes quotes in tool-related strings within the input message. This normalization is crucial for
+ * ensuring consistent JSON serialization and deserialization when working with tool calls. Specifically, the function
+ * helps avoid differences between `json::dump` and `json::parse` behavior, which can occur when single quotes are used
+ * incorrectly in tools. This also leads to missing initial tool tag tokens (<|tool_call|>) during inference.
+ *
+ * In the case of Phi-4, when tools are embedded within messages, Minja's handling of tools within the message body
+ * (instead of treating them as separate input keys/values) can lead to issues with the serialization process. For instance,
+ * `json::dump` might escape single quotes, but `json::parse` might not interpret them correctly, resulting in inconsistent
+ * output when processing tools as part of a message.
+ *
+ * To mitigate this, the function scans through the string and normalizes single quotes to double quotes in the context
+ * of tool calls. This ensures that tool keys and values are serialized consistently, and the tool call context remains
+ * intact. The function specifically looks for single quotes that are preceded or followed by specific characters
+ * (such as '{', ':', ' ', or ',') which typically indicate the presence of tool keys and values.
+ *
+ * Parameters:
+ * - input (const std::string&): The input string containing single quotes that may need to be normalized.
+ *
+ * Returns:
+ * - std::string: A new string where appropriate single quotes are replaced by double quotes.
+ *
+ * Example:
+ * - If the input string is "{'key': 'value'}", the function will return "{\"key\": \"value\"}".
```

Comment:
could a single quote within a quoted string be replaced with a double quote? is that the intended behavior?

`"some key": "some value that has a single quote: ' "`